src/core/server/saved_objects/migrationsv2/actions.ts

```diff
@@ -0,0 +1,811 @@
+        return TaskEither.right('create_index_succeeded');
+      } else {
+        // Otherwise, wait until the target index has a 'green' status.
+        return pipe(
+          waitForStatus(client, indexName, 'green'),
+          TaskEither.map(() => {
+            /** When the index status is 'green' we know that all shards were started */
+            return 'create_index_succeeded';
+          })
+        );
+      }
+    })
+  );
+};
+
+export interface UpdateAndPickupMappingsResponse {
+  taskId: string;
+}
+
+/**
+ * Updates an index's mappings and runs an update_by_query so that the mapping
+ * changes are "picked up". Returns a taskId to track progress.
+ *
+ * @param client
+ * @param index
+ * @param mappings
+ */
+export const updateAndPickupMappings = (
+  client: ElasticsearchClient,
+  index: string,
+  mappings: IndexMapping
+): TaskEither.TaskEither<RetryableEsClientError, UpdateAndPickupMappingsResponse> => {
+  const putMappingTask: TaskEither.TaskEither<
+    RetryableEsClientError,
+    { acknowledged: boolean }
+  > = () => {
+    return client.indices
+      .putMapping<Record<string, any>, IndexMapping>({
+        index,
+        timeout: '60s',
+        body: mappings,
+      })
+      .then((res) => {
+        // TODO do we need to check res.body.acknowledged?
+        return Either.right({ acknowledged: res.body.acknowledged });
+      })
+      .catch(catchRetryableEsClientErrors);
+  };
+
+  return pipe(
+    putMappingTask,
+    TaskEither.chain((res) => {
+      return updateByQuery(client, index);
+    })
+  );
+};
+export interface SearchResponse {
+  hits: SavedObjectsRawDoc[];
+}
+
+export const search = (
+  client: ElasticsearchClient,
+  index: string,
+  query: Record<string, unknown>
+): TaskEither.TaskEither<RetryableEsClientError, SearchResponse> => () => {
+  return client
+    .search<{
+      // when `filter_path` is specified, ES doesn't return empty arrays, so if
+      // there are no search results res.body.hits will be undefined.
+      hits?: {
+        hits?: SavedObjectsRawDoc[];
+      };
+    }>({
+      index,
+      // Optimize search performance by sorting by the "natural" index order
+      sort: ['_doc'],
+      // Return the _seq_no and _primary_term so we can use optimistic
+      // concurrency control for updates
+      seq_no_primary_term: true,
+      // Return batches of 100 documents. Smaller batches reduce the memory
+      // pressure on Elasticsearch and Kibana so are less likely to cause
+      // failures.
+      // TODO (profile/tune): How much smaller can we make this number before
+      // it starts impacting how long migrations take to perform?
+      size: 100,
+      body: {
+        query,
+      },
+      // Don't return partial results if timeouts or shard failures are
+      // encountered. Set explicitly to avoid users overriding the
+      // search.default_allow_partial_results cluster setting to false.
+      allow_partial_search_results: true,
```

Comment:
The comment is correct but the value isn't 😬 Because we interpret `0` results as there being no more document transformation work to be done, we want to be sure that the cluster always returns all results or an error.